### ISSUE TITLE
SQLite's DSL.extract() emulation should render an explicit CAST(STRFTIME(...) AS INT)

### ISSUE THREAD

**Issue by gotson** [jOOQ/jOOQ#11299]
### Expected behavior

When using `DSL.year()` on a `date` field, and a `in` condition, the comparison should work.

### Actual behavior

When using `DSL.year()` on a `date` field, and a `in` condition, the comparison fails, because `year` uses `strftime` to get the date part, and returns a string, but the other side of the comparison uses an `Int` instead.

### Steps to reproduce the problem

- MCVE: https://github.com/gotson/jOOQ-mcve/tree/sqlite-datepart

### Details

Adding a `cast` to `String` forces the other side of the comparison to String also, and succeeds.

Jooq statements with bind values:

```kotlin
ctx.selectDistinct(TEST.MYDATE)
      .from(TEST)
      .where(year(TEST.MYDATE).`in`(listOf("2000")))
      .fetchSet(TEST.MYDATE)
```

`select distinct test.mydate from test where strftime('%Y', test.mydate) in (2000)`

---

```kotlin
ctx.selectDistinct(TEST.MYDATE)
      .from(TEST)
      .where(year(TEST.MYDATE).`in`(listOf(2002)))
      .fetchSet(TEST.MYDATE)
```

`select distinct test.mydate from test where strftime('%Y', test.mydate) in (2002)`

---

```kotlin
ctx.selectDistinct(TEST.MYDATE)
      .from(TEST)
      .where(cast(year(TEST.MYDATE), String::class.java).`in`(listOf("2010")))
      .fetchSet(TEST.MYDATE)
```

`select distinct test.mydate from test where cast(strftime('%Y', test.mydate) as varchar) in ('2010')`

---

```kotlin
ctx.selectDistinct(TEST.MYDATE)
      .from(TEST)
      .where(cast(year(TEST.MYDATE), String::class.java).`in`(listOf(2015)))
      .fetchSet(TEST.MYDATE)
```

`select distinct test.mydate from test where cast(strftime('%Y', test.mydate) as varchar) in ('2015')`

### Versions

- jOOQ: 3.14.4
- Java: 8
- Database (include vendor): SQLite
- OS: Mac
- JDBC Driver (include name if inofficial driver): org.xerial:sqlite-jdbc:3.34.0

**Comment by lukaseder** [comment]
Thanks for your report.

Interesting case. The particular test that verifies correctness of this emulation just validates the projection, not usage in predicates. The `strftime` call should be wrapped in a cast back to `integer` as promised by the `DSL.extract()` function.

I'm surprised your `listOf("some-string")` calls compile, they shouldn't...?

**Comment by lukaseder** [comment]
There might be more emulations that suffer from incomplete testing. I'll think about how we can automate the discovery of such missing tests (functions being integration tested in projections but not in predicates) and fix potential issues in a new issue.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.6 (#11300).

As a workaround, for the time being, I suggest you cast your `year()` expressions to `SQLDataType.INTEGER` explicitly
